bob-access/src/authenticator/basic.rs

```diff
@@ -1,62 +1,148 @@
-use std::{collections::HashMap, net::IpAddr};
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    sync::{RwLock, Arc},
+    thread::sleep, 
+    time::Duration,
+};
 
 use crate::{credentials::{Credentials, CredentialsKind}, AuthenticationType, error::Error, permissions::Permissions};
 
 use super::{users_storage::UsersStorage, Authenticator};
 
 use sha2::{Digest, Sha512};
 
+use tokio::net::lookup_host;
+
+type NodesCredentials = HashMap<String, Credentials>;
+
 #[derive(Debug, Default, Clone)]
 pub struct Basic<Storage: UsersStorage> {
     users_storage: Storage,
-    nodes: HashMap<IpAddr, Vec<Credentials>>,
+    nodes: Arc<RwLock<NodesCredentials>>,
+    resolve_sleep_dur_ms: u64,
 }
 
 impl<Storage: UsersStorage> Basic<Storage> {
-    pub fn new(users_storage: Storage) -> Self {
+    pub fn new(users_storage: Storage, resolve_sleep_dur_ms: u64) -> Self {
         Self {
             users_storage,
-            nodes: HashMap::new(),
+            nodes: Arc::new(RwLock::new(HashMap::new())),
+            resolve_sleep_dur_ms,
         }
     }
 
+    fn node_creds_ok(creds: &HashMap<String, Credentials>) -> bool {
+        creds.values()
+            .all(|cred|
+                cred.ip().is_some() &&
+                cred.kind().map(|k| k.is_internode()) == Some(true))
+    }
+
+    fn unresolved_node_creds_ok(creds: &Vec<Credentials>) -> bool {
+        creds.iter()
+            .all(|cred|
+                cred.hostname().is_some() &&
+                cred.kind().map(|k| k.is_internode()) == Some(true))
+    }
+
     pub fn set_nodes_credentials(
         &mut self,
-        nodes: HashMap<IpAddr, Vec<Credentials>>,
+        nodes: HashMap<String, Credentials>,
+        unresolved: Vec<Credentials>,
```

Comment:
Why do you split nodes with `ip` and with host name in arguments. IMHO, this function should received them without distinction. Inside this function you can split them if it is required